### PR TITLE
FTP Camera in Animal Detection mode

### PR DESCRIPTION
**Pull request by stefanodellosa-personal**

### CODE REVIEW

File: src/domain/animal_detection_mode.py [hunk 1]
```diff
@@ -7,6 +7,9 @@
 from src.domain.camera import img_queue
 from src.domain.camera import Camera
 from src.domain.camera import cameras
+from src.domain.ftps_server import FTPsServer
+
+from PySide6.QtCore import QThread
```
Comment (cdimauro) [inline src/domain/animal_detection_mode.py:12]: from PySide6.QtCore import QThread

from src.domain.camera import Camera, cameras, img_queue
from src.domain.ftps_server import FTPsServer
from src.domain.operation_mode import OperationMode

File: src/domain/animal_detection_mode.py [hunk 2]
```diff
@@ -29,13 +31,16 @@ def run(self):
...
         for camera in cameras:
-            if not camera.enabled or camera.type != Camera.CameraTypes.USBCamera:
-                continue
-            else:
+            if camera.enabled and camera.type == Camera.CameraTypes.USBCamera:
                 # Create thread for motion detection
                 logger.debug("Instantiating thread for camera %s", camera.id)
                 camera.stop_thread = False
                 self.camera_thread.append(camera.run())
+            elif camera.enabled and camera.type == Camera.CameraTypes.FTPCamera and FTPsServer.ftps_server:
```
Comment (cdimauro) [inline src/domain/animal_detection_mode.py:39]: You've already checked camera.enabled on the if statement. Maybe you can reorganize the ifs like:

if camera.enabled:
    if camera.type == Camera.CameraTypes.USBCamera:
    [...]
    elif camera.type == Camera.CameraTypes.FTPCamera:

File: src/domain/animal_detection_mode.py [hunk 3]
```diff
@@ -29,13 +31,16 @@ def run(self):
...
+                logger.info("Instantiating FTPS server...")
+                self.init_ftp_server()
+            else:
+                continue
```
Comment (cdimauro) [inline src/domain/animal_detection_mode.py:43]: Redundant

File: src/domain/animal_detection_mode.py [hunk 4]
```diff
@@ -71,3 +76,30 @@ def check_for_termination_requests(self):
...
+        # Move operation mode in dedicated thread
+        FTPsServer.ftps_server.moveToThread(self.ftp_thread)
+
+        # Connect thread related signals and slots
+        self.ftp_thread.started.connect(FTPsServer.ftps_server.run)
+        FTPsServer.ftps_server.run_finished.connect(self.ftp_thread.quit)
+        FTPsServer.ftps_server.run_finished.connect(FTPsServer.ftps_server.deleteLater)
+        self.ftp_thread.finished.connect(self.ftp_thread.deleteLater)
+
+        # Start the thread
```
Comment (cdimauro) [inline src/domain/animal_detection_mode.py:104]: Redundant

File: src/domain/ftps_server.py [hunk 5]
```diff
@@ -4,21 +4,28 @@
 from pyftpdlib.authorizers import DummyAuthorizer
 from pyftpdlib.servers import ThreadedFTPServer
 from pyftpdlib.handlers import TLS_FTPHandler
-
 from PySide6.QtCore import QObject, Signal
+from src.domain.camera import img_queue
```
Comment (cdimauro) [inline src/domain/ftps_server.py:8]: from pyftpdlib.authorizers import DummyAuthorizer
from pyftpdlib.handlers import TLS_FTPHandler
from pyftpdlib.servers import ThreadedFTPServer
from PySide6.QtCore import QObject, Signal

from src.domain.camera import img_queue

File: src/domain/ftps_server.py [hunk 6]
```diff
@@ -4,21 +4,28 @@
...
 
 logger = logging.getLogger(__name__)
 
+class TLS_FTP_WADAS_Handler(TLS_FTPHandler):
+
+    def on_connect(self):
+        logger.info(f"Connected remote camera from {self.remote_ip}:{self.remote_port}")
```
Comment (cdimauro) [inline src/domain/ftps_server.py:15]: Never use f-strings on logger methods.

logger.info("Connected remote camera from %s: %s", self.remote_ip, self.remote_port)

File: src/domain/ftps_server.py [hunk 7]
```diff
@@ -57,7 +64,7 @@ def add_user(self, username, password, directory):
         self.authorizer.add_user(username, password, directory, perm='elradfmwMT')
```
Comment (cdimauro) [inline src/domain/ftps_server.py:64]: Security hole:

perm='elradfmwMT'

? :)

File: src/domain/operation_mode.py [hunk 8]
```diff
@@ -13,6 +13,7 @@
 
 from PySide6.QtCore import QObject, Signal
 from src.domain.ai_model import AiModel
+from src.domain.ftps_server import FTPsServer
```
Comment (cdimauro) [inline src/domain/operation_mode.py:16]: from email.mime.image import MIMEImage
from email.mime.multipart import MIMEMultipart
from email.mime.text import MIMEText
import logging
import os
import smtplib

import keyring
from PySide6.QtCore import QObject, Signal
import ssl

from src.domain.ai_model import AiModel
from src.domain.ftps_server import FTPsServer

File: src/domain/animal_detection_mode.py (hunk 4 above)
Comment (stefanodellosa-personal) [reply src/domain/animal_detection_mode.py:104]: They are actually different slots.. finished vs run_finished

File: src/domain/ftps_server.py (hunk 6 above)
Comment (stefanodellosa-personal) [reply src/domain/ftps_server.py:15]: good catch!

File: src/domain/ftps_server.py (hunk 7 above)
Comment (stefanodellosa-personal) [reply src/domain/ftps_server.py:64]: removed unnecessary permissions

File: src/domain/ftps_server.py [hunk 9]
```diff
@@ -85,6 +86,23 @@ def run(self):
             self.server.serve_forever()
         self.run_finished.emit()
 
+    #TODO: remove once test is completed
+    def run_op_mode(self):
+        """ TEST Method to create new thread for FTPS Server class."""
+
+        if self.server:
+            thread = threading.Thread(target=self.server.serve_forever())
```
Comment (cdimauro) [inline src/domain/ftps_server.py:94]: is self.server.serve_forever() giving back the callback that Thread should use? If not and you want Thread to use it as callback, then remove the ().

File: src/domain/ftps_server.py [hunk 10]
```diff
@@ -1,14 +1,15 @@
 """FTPS server module"""
 
 import logging
+import threading
 from pyftpdlib.authorizers import DummyAuthorizer
```
Comment (cdimauro) [inline src/domain/ftps_server.py:5]: Add blank line to separate imports.

File: src/domain/ftps_server.py (hunk 9 above)
Comment (stefanodellosa-personal) [reply src/domain/ftps_server.py]: This was actually the issue!!!!!

**Pull request merged**